worlds/__init__.py

```diff
@@ -95,27 +100,96 @@ def load(self) -> bool:
-    "games": {world_name: world.get_data_package_data() for world_name, world in AutoWorldRegister.world_types.items()},
-}
+def scan_worlds() -> List[WorldSource]:
+    # find potential world containers, currently folders and zip-importable .apworld's
+    sources: List[WorldSource] = []
+    for folder in (folder for folder in (user_folder, local_folder) if folder):
+        relative = folder == local_folder
+        for entry in os.scandir(folder):
+            # prevent loading of __pycache__ and allow _* for non-world folders, disable files/folders starting with "."
+            if not entry.name.startswith(("_", ".")):
+                file_name = entry.name if relative else os.path.join(folder, entry.name)
+                if entry.is_dir():
+                    sources.append(WorldSource(file_name, relative=relative))
+                elif entry.is_file() and entry.name.endswith(".apworld"):
+                    sources.append(WorldSource(file_name, is_zip=True, relative=relative))
+    sources.sort()
+    return sources
+
+
+def get_world_paths_from_name(name: str) -> Tuple[str, str]:
+    for source, data in world_data.items():
+        if data["game"] == name:
+            return source, data["path"]
+    else:
+        raise ModuleNotFoundError(f"No game found for {name}")
+
+
+def get_worlds_info() -> Tuple[Dict[str, Dict[str, str]], bool]:
+    should_update = False
+    world_data = {}
+    try:
+        cached_time = os.path.getmtime(cached_worlds_path)
+        world_data = orjson.loads(pkgutil.get_data(__name__, cached_worlds_path))
+        for world in world_sources:
+            if os.path.getmtime(world.resolved_path) > cached_time:
+                should_update = True
+                break
+    except (FileNotFoundError, orjson.JSONDecodeError):
+        should_update = True
+
+    return world_data, should_update
+
+
+def load_all_worlds() -> DataPackage:
+    # import all submodules to trigger AutoWorldRegister
+    games: Dict[str, GamesPackage] = {}
+    json_data: Dict[str, Dict[str, str]] = {}
+    for world_source in world_sources:
+        world_source.load()
+
+    for world_name, world in AutoWorldRegister.world_types.items():
+        games[world_name] = world.get_data_package_data()
+        # GitHub runners won't allow me to write to their cache
+        if "runner" in cached_worlds_path:
+            continue
+        json_name = [name for name in world.__module__.split(".") if "world" not in name][0]
+        if world.zip_path:
+            json_name += ".apworld"
+        world_json = {"game": world.game, "path": world.__file__,
+                      "settings": str(world.settings_key) if hasattr(world, "settings") else None}
+        json_data[json_name] = world_json
+
+    if "runner" not in cached_worlds_path:
+        with open(cached_worlds_path, "wb") as f:
+            f.write(orjson.dumps(json_data))
+    return DataPackage(games=games)
+
+
+def load_worlds(games: Iterable[str]) -> DataPackage:
```

Comment:
```suggestion
def load_worlds(games: Union[Iterable[str], str]) -> DataPackage:
    if isinstance(games, str):
        games = (games,)

```
With collections of `str`, this is a good pattern to use if it makes any sense at all.

Notice that the type checker won't see any problem with:
`load_worlds("Messenger")`
because a `str` is an `Iterable[str]`

So since we can't get the type checker to help us make sure a `str` isn't passed, I think it's better if we support passing a `str`

Reply:
I'm not sure if this is worth supporting or not. I was supporting it before, but the only cases where this is ever viable are generate (generating single player game) and launcher (component lazy loading). For the launcher, you could just easily wrap it, but for generate I think you'd want to always parse the necessary games that it would support multiple games, since that's "default". So I guess the question is if this is worth it for something that'd only be an additional 2 characters in the launcher vs having an if check here every time a world/s get loaded.

Reply:
> an additional 2 characters in the launcher vs having an if check here every time a world/s get loaded

an additional 2 characters in the launcher and the type checker can't help us make sure the right type is passed
vs
having an if check here every time a world/s get loaded and the type checker can help us make sure the right type is passed

I think it's worth it even if the launcher and generate don't end up using it, just for the type safety.